### PR TITLE
[stable/elasticsearch] Add option to provide custom start/stop hooks

### PR DESCRIPTION
**Pull request by melifaro**
#### What this PR does / why we need it:
Currently start/stop hooks are hardcoded in ConfigMap, and will wait for the all shards to be relocated from the nodes.
It takes very long time for the huge ES data nodes (logs clusters, for example). 
This PR provides the option to customize it using values and tpl function.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by melifaro** [comment]
I'm done here with the PR, also added custom hooks option for client deployment (we use them to drain connections before terminating the pods).
Can I please have a review?

**Comment by melifaro** [comment]
@desaintmartin can you please have a look?

File: stable/elasticsearch/values.yaml [hunk 1]
```diff
@@ -214,6 +220,13 @@ data:
   hooks:  # post-start and pre-stop hooks
     drain:  # drain the node before stopping it and re-integrate it into the cluster after start
       enabled: true
+    # If you need modify the standart drain hooks, set hooks.drain.enabled to false
```
Comment (legal90) [inline stable/elasticsearch/values.yaml:223]: That behavior is not intuitive and might look confusing for users. 
"Why should I disable the drain hook if I want the custom one to be executed?"

I suggest to add `data.hooks.drain.pre-stop` and `data.hooks.drain.pre-start` variables with an empty default value. And then render the default body of these scripts in `_helpers.yaml`.  That will allow to keep the backward compatibility with a minimal footprint on `values.yaml`

File: stable/elasticsearch/values.yaml [hunk 2]
```diff
@@ -118,6 +118,12 @@ client:
     #  - secretName: chart-example-tls
     #    hosts:
     #      - chart-example.local
+  hooks:
+    enabled: false
+    # enabled: false
+    # start: |
+    # stop: |
```
Comment (legal90) [inline stable/elasticsearch/values.yaml:125]: I would suggest to keep the consistency with `data.` section, so:
```yaml
  hooks:
    drain:
      enabled: false
      # enabled: false
      # pre-start: |
      # pre-stop: |
```

**Review by legal90 — COMMENTED** [review]
Thank you for this PR! 👍
I was struggling with the same issue. I want the pre-stop hook just exclude the node from the the allocation (in order to do a quick rolling update, without actual draining the data node.

However, I have some notes regarding the implementation of custom hooks. I will try to create a new PR soon showing my vision on the possible solution.

Anyway, it's great that you initiated this PR.  Great job!

File: stable/elasticsearch/templates/client-deployment.yaml [hunk 3]
```diff
@@ -169,3 +169,18 @@ spec:
         secret:
           secretName: {{ .Values.cluster.keystoreSecret }}
 {{- end }}
+{{- if .Values.client.hooks.enabled }}
```
Comment (tetianakravchenko) [inline stable/elasticsearch/templates/client-deployment.yaml:172]: in this case, `terminationGracePeriodSeconds` also should be added for client

**Comment by legal90** [comment]
Hi guys!
I rework this PR a little bit by making hooks more generic, fixing a README.md markup and adding `terminationGracePeriodSeconds` as @tetianakravchenko has suggested.

Please, take a look: https://github.com/helm/charts/pull/12206 
Suggestions are welcome!

**Pull request closed without merging**